DL Caching on the client

gmalette opened the pull request:
This PR adds caching to DL both on the server side and on the client side. The branch itself is based on #4 (server-side), in case we want to merge server and client separarely. I would strongly encourage reading #4 for a description of the changes for `fs.GetCache()`.

## `client.GetCache()`

The client has a new `GetCache()` function which allows populating the provided cache directory with the contents returned by the server. The hierarchy changed from the original proposal:

```
<cache root> /
    - .lock
    - versions
    - tmp/
    - objects/
        <cache key> /
            <packed content>
```

The `.lock` file is only present when an ongoing `GetCache()` request is performed on this cache root. The goal is to avoid two processes racing to populate the same cache root. The file is created with `O_EXCL`, meaning that only one process can create this file; others will get an error and abort.

The `versions` file is a plain-text file containing the versions available.

The `tmp` directory is where the files are extracted. This is done to prevent an extraction failure from corrupting the cached objects. The files are then moved to their final destination, in the `objects` directory.

**NOTE**: If the process crashes while fetching the cache, the `.lock` file will not be cleaned up and it will not be possible for other processes to populate the cache root. This was chosen in order to avoid undefined behaviour within the cache.

## `fs.GetCompress` 

This PR mainly changes `fs.GetCompress` and its dependencies to allow caching packages. 

- A new CTE is added to the `updated_objects`, allowing a `LEFT JOIN` on the cached versions sent by the client
- The performed DB queries will no longer return the `bytes` column for a cached entry
- I'm abusing the tarfile content to send the cache hash. On the client-side the hash is hex-encoded to produce the path in the cache folder, to which we append the file name and symlink there.

gmalette committed:
[cache] Introduce cache computation for impactful node_modules

gmalette committed:
[cache] Test GetCache without available cache_versions

gmalette committed:
[cache]  Make the query a little nicer

gmalette committed:
[cache] Client code

gmalette committed:
[cache] Move Cached flag from pb.Object to a new TarObject

gmalette committed:
[cache] Add tests

angelini commented on internal/db/query.go:
```diff
@@ -165,12 +166,12 @@ func GetObjects(ctx context.Context, tx pgx.Tx, packManager *PackManager, projec
 
 	if err != nil {
-		return nil, fmt.Errorf("getObjects query, project %v vrange %v: %w", project, vrange, err)
+		return nil, fmt.Errorf("1getObjects query, project %v vrange %v: %w", project, vrange, err)
```
I think this change was a typo.

angelini commented on internal/db/query.go:
```diff
@@ -189,16 +190,21 @@ func GetObjects(ctx context.Context, tx pgx.Tx, packManager *PackManager, projec
 		var h1, h2 []byte
 
-		err := rows.Scan(&path, &mode, &size, &encoded, &packed, &deleted, &h1, &h2)
+		err := rows.Scan(&path, &mode, &size, &cache_h1, &cache_h2, &encoded, &packed, &deleted, &h1, &h2)
```
As mentioned in the other PR, we can use the `Hash` type here.

```
var hash Hash
rows.Scan(&hash.H1, &hash.H2)
```

angelini commented on internal/db/query.go:
```diff
@@ -250,31 +256,39 @@ func GetTars(ctx context.Context, tx pgx.Tx, project int64, vrange VersionRange,
+		if cache_h1 != nil {
+			cache_hash = make([]byte, 32)
+			copy(cache_hash[0:16], cache_h1)
+			copy(cache_hash[16:], cache_h2)
```
This can be replaced by the `.Bytes()` method on `Hash`

angelini commented on internal/db/query.go:
```diff
@@ -332,3 +402,32 @@ func (p *PackManager) IsPathPacked(path string) *string {
+				AND stop_version IS NULL
+			GROUP BY hash
+			ORDER BY count DESC
+			LIMIT 100
```
As mentioned in the other PR, should this become a function var and maybe even an argument to the `CreateCache` RPC.

angelini commented on internal/db/queryBuilder.go:
```diff
@@ -118,29 +122,39 @@ func (qb *queryBuilder) removedObjectsCTE() string {
-		), updated_objects AS (
+		), cached_object_hashes AS (
+	%s
+	), updated_objects AS (
```
Nit: There's some formatting misalignment here that shows up in Github.

angelini commented on internal/db/queryBuilder.go:
```diff
@@ -149,18 +163,18 @@ func (qb *queryBuilder) queryWithRemovals() string {
 
 	selectStatement := `
-		SELECT path, mode, size, bytes, packed, deleted, (hash).h1, (hash).h2
+		SELECT path, mode, size, (cache_hash).h1 cache_h1, (cache_hash).h2 cache_h2, bytes, packed, deleted, (hash).h1, (hash).h2
```
Is the `cache_hash` different than the `hash`?

It seems like we may be returning the same values twice. Could we always return `hash` and then include a `is_hashed` boolean?

gmalette replied:
Yeah possibly. I don't know why `hash` isnt't always returned; it was added during one of the rebases. I'll try to find out.

angelini commented on internal/files/writer.go:
```diff
@@ -2,13 +2,16 @@ package files
 	"os"
 	"path/filepath"
 
+	"github.com/gadget-inc/dateilager/internal/pb"
```
Is this importing the same module twice?

gmalette replied:
which other instance are you referring to?

angelini replied:
Ah that was a mistake on my part, `internal/pb` and `internal/db` look very similar 🤦 

It looks like you likely don't have the Go imports sorter installed. It should have sorted these alphabetically and removed the newline below here.

gmalette replied:
I think it's because it's not imported in the `.golangci.yml` file. Since I'm importing that file, my editor doesn't let me enable extra ones 🤷🏼‍♂️

angelini commented on internal/pb/fs.go:
```diff
@@ -9,7 +9,12 @@ import (
-func tarTypeFromMode(mode fs.FileMode) byte {
+const (
+	TarDeleted = 'D'
+	TarCached  = 'C'
```
Sounds like we're good to use `'C'`

<img width="638" alt="image" src="https://user-images.githubusercontent.com/515110/194598192-f9a07072-9338-48d0-9497-f943e4f420c3.png">

angelini commented on pkg/api/fs.go:
```diff
@@ -862,3 +862,46 @@ func (f *Fs) GcContents(ctx context.Context, req *pb.GcContentsRequest) (*pb.GcC
+		if err == io.EOF {
+			break
+		}
+		// if err == db.SKIP {
```
As mention in the other PR, we don't need to handle this here since we always get back then entire cache.

angelini commented on test/db_test.go:
```diff
@@ -0,0 +1,156 @@
+
+		hash := make([]byte, 32)
+		copy(hash[0:16], h1)
+		copy(hash[16:32], h2)
```
This can also make use of the `Hash.Bytes()` method

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+	if err != nil {
+		return -1, err
+	}
+	os.RemoveAll(tmpObjectDir)
```
You can set this up as a `defer` right after creating the dir, which will ensure that gets removed even in error cases.

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+		return -1, fmt.Errorf("fs.GetCache failed to update the versions file: %w", err)
+	}
+
+	err = cleanupCacheLockFile(lockFile)
```
So we kind of want to make sure this happens no matter what.

What do you think if we made it so that this function cannot return an error and then add it as a `defer` right after we create the lock.

Trying to avoid cases where we can potentially leave a directory in a bad state that cannot be recovered from.

gmalette replied:
I very much disagree, especially since we currently do not have any recovery mechanisms. 

I'd very much rather have an irrecoverable cache that cannot be used than an inconsistent cache that *may* be recovered from.

In the former case, we're going to be paying more for the transfer of packed files because the cache will be useless (and irrecoverable).

In the latter case, we'd have to deal with multiple, unknown failure modes. We'd have to reason about what all those are before trying to recover from a failure.

Sidenote: I'm starting to think it's harmful to describe the current solution as a "cache", at least from the client's perspective. If there's a cache miss, the application will fail to boot; it will not run in a degraded performance kind of way.

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+	})
+
+	for i := 0; i < parallelWorkerCount(); i++ {
+		// create the attribute here when `i` is different
```
🙌 

There's actually a proposal to fix this in upcoming Go versions where the beginning of every loop will have an implicit `i := i`

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+
+					if fileExists(tempDest) {
+						err := os.RemoveAll(tempDest)
+						return fmt.Errorf("temporary cache folder exists for %s and couldn't be removed: %w", tempDest, err)
```
Should this be `if err != nil`?

And should we call `cancel()` in this case?

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+
+	for _, availableVersion := range availableVersions {
+		if version == availableVersion {
+			return availableVersions[len(availableVersions)-1], nil
```
It's not clear to me why we're returning the last `availableVersion` in this case.

gmalette replied:
You mean, as opposed to the one returned by the server? Since we're skipping the unpacking, I figure that returning the last one that was received makes sense, but now that I say it out loud perhaps not. 

I'll change it to "the one that the server returned" but TBF I've been struggling with the return valud of `GetCache` and I think it should probably either be `void` or `[]int64`

angelini commented on pkg/client/client.go:
```diff
@@ -557,6 +562,177 @@ func (c *Client) Reset(ctx context.Context, state string) error {
+					tarReader := db.NewTarReader(response.Bytes)
+					hashHex := hex.EncodeToString(response.Hash)
+					tempDest := filepath.Join(tmpObjectDir, hashHex)
+					finalDest := filepath.Join(objectDir, hashHex)
```
Why do we need both the `temp` and `final` destinations? Since we're taking a lock is it not safe to write directly into the final destination?

gmalette replied:
Well, the thing is that we don't know that the extraction will succeed. If we extract directly to the final destination we risk creating an invalid file structure. At present, even if we wrote a bad file, we would not use it, but I can't guarantee that this will hold, and I haven't written tests to prove it either.

The easiest way to avoid that situation is to extract somewhere else and avoid moving if the extraction fails.

angelini replied:
Ahh got it.

Our two options are to work in a temp directory (as you are doing) or do a full cleanup (`rm -rf`) at the beginning of Cache Creation to potentially drop any failed previous attempt.

Is that second option possible / simpler, or not something we can do in this case.

gmalette replied:
A worker could have an in-flight request using existing cache versions. We can *never* delete what has been made available, unless we implement synchronization

angelini replied:
Yeah 100%, we should never delete something once it has been made available.

But if it was never made available then it should be safe to delete?

gmalette replied:
yes, if and only if it was never made available, we can delete it, the trick is in knowing

angelini commented on internal/db/query.go:
```diff
@@ -291,6 +305,62 @@ func GetTars(ctx context.Context, tx pgx.Tx, project int64, vrange VersionRange,
+
+		hash := make([]byte, 32)
+		copy(hash[0:16], h1)
+		copy(hash[16:], h2)
```
Another case where the `Hash` type will add better type info to this function and allow you to not have to repeat this code.

gmalette committed:
[cache] Change CreateCache count to a parameter instead of fixed 100

gmalette committed:
[cache] Formatting fixes

gmalette committed:
[cache] Use defer to remove temp dir when unpacking cache files

gmalette committed:
[cache] Replace usage of []byte for hashes with Hash type

gmalette committed:
[cache] Reuse the existing cache SQL selection instead of duplicating

gmalette committed:
[cache] Return the version returned by the server, not the last downloaded one

gmalette committed:
[cache] Fix err handling while extracting the cache

gmalette committed:
[cache] Make sure that GetCache Project auths

gmalette committed:
[cache] Always cleanup tmpObjectDir and lock file, even if GetCache is unsuccessful

Since objects are unpacked into a tmpObjectDir, the main Object dir remains clean in the event of failures.

Also, since the versions file is only written to when the entire process succeeds, the workers will not try
to use a partially extracted GetCache stream.

gmalette committed:
[cache] Remove returned error from cleanupCacheLockFile since it's deferred

gmalette committed:
[cache] fix when files exist

angelini committed:
[cache] review comments and GetCompress test
